flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java

```diff
@@ -0,0 +1,299 @@
+
+    private final FileSystem fs;
+
+    private final Path targetFile;
+
+    private final Path tempFile;
+
+    private final FSDataOutputStream out;
+
+    // Not final to override in tests
+    public static int minBufferLength = 2097152;
+
+    // init to 0. When ever recovery is done add this to the pos.
+    private long initialFileSize = 0;
+
+    public AzureBlobFsRecoverableDataOutputStream(FileSystem fs, Path targetFile, Path tempFile)
+            throws IOException {
+        this.fs = checkNotNull(fs);
+        this.targetFile = checkNotNull(targetFile);
+        LOG.debug("The targetFile is {}", targetFile.getName());
+        this.tempFile = checkNotNull(tempFile);
+        LOG.debug("The tempFile is {}", tempFile.getName());
+        this.out = fs.create(tempFile);
+    }
+
+    public AzureBlobFsRecoverableDataOutputStream(FileSystem fs, HadoopFsRecoverable recoverable)
+            throws IOException {
+        this.fs = checkNotNull(fs);
+        this.targetFile = checkNotNull(recoverable.targetFile());
+        this.tempFile = checkNotNull(recoverable.tempFile());
+        long len = fs.getFileStatus(tempFile).getLen();
+        LOG.info("The recoverable offset is {} and the file len is {}", recoverable.offset(), len);
+        // Happens when we recover from a previously committed offset. Otherwise this is not
+        // really needed
+        if (len > recoverable.offset()) {
+            truncate(fs, recoverable);
+        } else if (len < recoverable.offset()) {
+            LOG.error(
+                    "Temp file length {} is less than the expected recoverable offset {}",
+                    len,
+                    recoverable.offset());
+            throw new IOException(
+                    "Unable to create recoverable outputstream as length of file "
+                            + len
+                            + " is less than "
+                            + "recoverable offset "
+                            + recoverable.offset());
+        }
+        // In ABFS when we try to append we don't account for the initial file size like we do in
+        // DFS.
+        // So we explicitly store this and when we do a persist call we make use of it.
+        initialFileSize = fs.getFileStatus(tempFile).getLen();
+        out = fs.append(tempFile);
+        LOG.debug("Created a new OS for appending {}", tempFile);
+    }
+
+    private void truncate(FileSystem fs, HadoopFsRecoverable recoverable) throws IOException {
+        Path renameTempPath = new Path(tempFile.toString() + ".rename");
+        try {
+            LOG.info(
+                    "Creating the temp rename file {} for truncating the tempFile {}",
+                    renameTempPath,
+                    tempFile);
+            FSDataOutputStream fsDataOutputStream = fs.create(renameTempPath);
+            LOG.info("Opening the tempFile {} for truncate", tempFile);
+            FSDataInputStream fsDis = fs.open(tempFile);
```

Comment:
This means that, despite the src and dest files are on the same remote file system, all the data being copied have to go through the local system where this piece of code is executed. I wonder if there's a better way doing this. (Maybe not, just trying to double-check on this.)

Reply:
I meant the way we write the truncated content to the renamed temp file. We first read the data from the temp file into the memory, then write the data from memory to the renamed temp file. Here the memory is where the Flink file sink operator is executed, while both temp file and renamed temp file are on the remote ABFS. Data transmissions between the memory and the ABFS can be expensive, at least more expensive than data transmissions within ABFS. So I just wonder if there's anyway to copy data directly from the temp file to the renamed temp file, without going through the local memory.

Reply:
Got your question. I checked with the storage driver team (hadoop based storage driver), they said currently there is no direct truncate support. So the manual truncation is the option over here. A generic question though @xintongsong - In case of ORC/Parquet format does the file recovery flow take care of the truncate or those formats already are aware of the last committed offset? This is a question out of this ABFS support PR (in general)

Reply:
Bulk-encoded formats (ORC, Parquet, etc.) do not need truncation, because in-progress files are guaranteed to be finished / flushed on checkpoints. Truncation is only needed when an in-progress file can live through multiple checkpoints, which is only supported for row-encoded formats.

Reply:
Thanks @xintongsong  for this info.

Reply:
Can we resolve this conversation also @xintongsong - How ever after our tests I can update here before we want to commit this - if that is fine with you.

Reply:
Sounds like there's no off-the-shelf solution for this at the moment. I think we should at least understand the performance impact and mention it in the user documentation.

Let's leave this conversation unresolved as a reminder for now. This won't block the PR review.